File: examples2d/joints2.rs
use rapier_testbed2d::Testbed;
use rapier2d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Enable/disable softness.
     */
    let settings = testbed.example_settings_mut();
    let variable_softness = settings.get_or_set_bool("Variable softness", false);

    /*
     * Create the balls
     */
    // Build the rigid body.
    // NOTE: a smaller radius (e.g. 0.1) breaks Box2D so
    // in order to be able to compare rapier with Box2D,
    // we set it to 0.4.
    let rad = 0.4;
    let numi = 10; // Num vertical nodes.
    let numk = 10; // Num horizontal nodes.
    let shift = 1.0;

    let mut body_handles = Vec::new();

    for k in 0..numk {
        for i in 0..numi {
            let fk = k as f32;
            let fi = i as f32;

            let status = if i == 0 && k == 0 {
                RigidBodyType::Fixed
            } else {
                RigidBodyType::Dynamic
            };

            let rigid_body =
                RigidBodyBuilder::new(status).translation(Vector::new(fk * shift, -fi * shift));
            let collider = ColliderBuilder::ball(rad);
            let (child_handle, _) = world.insert(rigid_body, collider);

            let softness = if variable_softness {
                // If variable softness is enabled, joints closer to the fixed body are softer.
                SpringCoefficients {
                    natural_frequency: 5.0 * (i.max(k) + 1) as f32,
                    damping_ratio: 0.1 * (i.max(k) + 1) as f32,
                }
            } else {
                SpringCoefficients::joint_defaults()
            };

            // Vertical joint.
            if i > 0 {
                let parent_handle = *body_handles.last().unwrap();
                let joint = RevoluteJointBuilder::new()
                    .local_anchor2(Vector::new(0.0, shift))
                    .softness(softness);
                world.insert_impulse_joint(parent_handle, child_handle, joint);
            }

            // Horizontal joint.
            if k > 0 {
                let parent_index = body_handles.len() - numi;
                let parent_handle = body_handles[parent_index];
                let joint = RevoluteJointBuilder::new()
                    .local_anchor2(Vector::new(-shift, 0.0))
                    .softness(softness);
                world.insert_impulse_joint(parent_handle, child_handle, joint);
            }

            body_handles.push(child_handle);
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec2::new(numk as f32 * rad, numi as f32 * -rad), 20.0);
}

File: python/rapier-testbed/rapier_testbed/examples2/joints2.py
"""Port of examples2d/joints2.rs."""
from __future__ import annotations

import rapier2d as rp
from .._registry import register

CATEGORY = "Joints"
NAME = "Joints"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    rad = 0.4
    numi = 10
    numk = 10
    shift = 1.0

    body_handles = []

    for k in range(numk):
        for i in range(numi):
            if i == 0 and k == 0:
                rb_builder = rp.RigidBody.fixed()
            else:
                rb_builder = rp.RigidBody.dynamic()
            rb = rb_builder.translation((k * shift, -i * shift))
            child_handle = bodies.insert(rb)
            colliders.insert_with_parent(rp.Collider.ball(rad), child_handle, bodies)

            # Vertical joint.
            if i > 0:
                parent_handle = body_handles[-1]
                joint = rp.RevoluteJoint.builder().local_anchor2((0.0, shift))
                impulse_joints.insert(parent_handle, child_handle, joint, wake_up=True)

            # Horizontal joint.
            if k > 0:
                parent_index = len(body_handles) - numi
                parent_handle = body_handles[parent_index]
                joint = rp.RevoluteJoint.builder().local_anchor2((-shift, 0.0))
                impulse_joints.insert(parent_handle, child_handle, joint, wake_up=True)

            body_handles.append(child_handle)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(numk * rad, numi * -rad), zoom=20.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
